reactive/kotlinx-coroutines-rx2/src/RxScheduler.kt

```
@@ -4,16 +4,129 @@
+ */
+public fun CoroutineDispatcher.asScheduler(): Scheduler =
+    if (this is SchedulerCoroutineDispatcher) {
+        scheduler
+    } else {
+        DispatcherScheduler(this)
+    }
+
+private class DispatcherScheduler(internal val dispatcher: CoroutineDispatcher) : Scheduler() {
+
+    private val job = SupervisorJob()
+    private val scope = CoroutineScope(job + dispatcher)
+
+    override fun scheduleDirect(block: Runnable): Disposable {
+        if (scope.isActive) {
+            return scope.launch {
+                block.runWithRx()
+            }.asDisposable()
+        }
+
+        return Disposables.disposed()
+    }
+
+    override fun scheduleDirect(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+        if (delay <= 0) {
+            return scheduleDirect(block)
+        }
+
+        if (scope.isActive) {
+            return scope.launch {
+                delay(unit.toMillis(delay))
+                block.runWithRx()
+            }.asDisposable()
+        }
+
+        return Disposables.disposed()
+    }
+
+    private fun Runnable.runWithRx() = RxJavaPlugins.onSchedule(this).run()
+
+    override fun createWorker(): Worker =
+        DispatcherWorker(dispatcher, job)
+
+    override fun shutdown() {
+        scope.cancel()
+    }
+
+    private class DispatcherWorker(dispatcher: CoroutineDispatcher, parentJob: Job) : Worker() {
+
+        private val workerScope = CoroutineScope(SupervisorJob(parentJob) + dispatcher)
+        private val blockChannel = Channel<Job>(Channel.UNLIMITED)
+        private var queueProcessingJob: Job? = null
+
+        override fun isDisposed(): Boolean = !workerScope.isActive
+
+        override fun schedule(block: Runnable): Disposable {
+            startProcessingQueue()
+            if (workerScope.isActive) {
+                val job = workerScope.launch(start = CoroutineStart.LAZY) {
+                    block.run()
+                }
+                blockChannel.offer(job)
```

If the channel has been closed, this will throw an exception, which i don't think should be thrown from this method directly - maybe reported to `RxPlugins.onError`?

---

I expect the channel to be closed at the same time as ```workerScope``` is set to inactive. So it should not even reach here if the channel has been closed. If it does I think it's better to let the exception be thrown so we can immediately know and fix.